### PR TITLE
Loads big models in bin file format faster

### LINKED ISSUE #14: Model load takes a very long time!
I am using the `Word2VecModel.fromBinFile(modelFile)` API to open the default word2vec model, `GoogleNews-vectors-negative300.bin` (https://code.google.com/p/word2vec/). The model is ~3.6 gigs. My program has been stuck loading this model for more than an hour now. Heap usage is at 11 gigs. 

What is the expected load time for a model this size? Why is it taking 11 gigs for a 3.6 gig model? 
Would converting the model to the Thrift format be a solution?

Appreciate your help!

Regards,
Sumithra

### PR DESCRIPTION
**Pull request by dirkgr**
This fixes #14.

### CODE REVIEW

File: src/main/java/com/medallia/word2vec/Word2VecModel.java [hunk 1]
```diff
@@ -76,69 +79,113 @@ public static Word2VecModel fromTextFile(File file) throws IOException {
...
+	/**
+	 * @return {@link Word2VecModel} created from the binary representation output
+	 * by the open source C version of word2vec using the given byte order.
+	 */
+	public static Word2VecModel fromBinFile(File file, ByteOrder byteOrder)
+			throws IOException {
+
+		try (FileInputStream fis = new FileInputStream(file);) {
+			final FileChannel channel = fis.getChannel();
+			final long oneGB = 1024 * 1024 * 1024;
```
Comment (wko27) [inline src/main/java/com/medallia/word2vec/Word2VecModel.java]: Can we extract this to a private static final and rename to ONE_GB?

File: src/main/java/com/medallia/word2vec/Word2VecModel.java [hunk 2]
```diff
@@ -76,69 +79,113 @@ public static Word2VecModel fromTextFile(File file) throws IOException {
...
+	}
+
+	/**
+	 * @return {@link Word2VecModel} created from the binary representation output
+	 * by the open source C version of word2vec using the given byte order.
+	 */
+	public static Word2VecModel fromBinFile(File file, ByteOrder byteOrder)
+			throws IOException {
+
+		try (FileInputStream fis = new FileInputStream(file);) {
```
Comment (wko27) [inline src/main/java/com/medallia/word2vec/Word2VecModel.java]: bit of a nit, but the semicolon on the end is unnecessary

File: src/main/java/com/medallia/word2vec/Word2VecModel.java [hunk 3]
```diff
@@ -17,29 +17,32 @@
...
  * <ul>
  * <li> {@link #trainer()}
  * <li> {@link #fromThrift(Word2VecModelThrift)}
  * </ul>
- * 
+ *
  * @see {@link #forSearch()}
  */
 public class Word2VecModel {
+	static Logger logger = Logger.getLogger(Word2VecModel.class);
```
Comment (wko27) [inline src/main/java/com/medallia/word2vec/Word2VecModel.java]: We have a ProfilingTimer abstraction which lets the user decide if he wants that info logged or e.g. displayed on a UI.  See notes below ...

File: src/main/java/com/medallia/word2vec/Word2VecModel.java [hunk 4]
```diff
@@ -76,69 +79,113 @@ public static Word2VecModel fromTextFile(File file) throws IOException {
...
+	public static Word2VecModel fromBinFile(File file)
+			throws IOException {
+		return fromBinFile(file, ByteOrder.LITTLE_ENDIAN);
+	}
+
+	/**
+	 * @return {@link Word2VecModel} created from the binary representation output
+	 * by the open source C version of word2vec using the given byte order.
+	 */
+	public static Word2VecModel fromBinFile(File file, ByteOrder byteOrder)
```
Comment (wko27) [inline src/main/java/com/medallia/word2vec/Word2VecModel.java]: Can we add a param for a ProfilingTimer and use that to create a subtask per 1 gb read?

I'd also create a version of fromBinFile which doesn't take the profiling timer and forwards with ProfilingTimer.NONE

File: src/main/java/com/medallia/word2vec/Word2VecModel.java [hunk 5]
```diff
@@ -76,69 +79,113 @@ public static Word2VecModel fromTextFile(File file) throws IOException {
...
+			final FileChannel channel = fis.getChannel();
+			final long oneGB = 1024 * 1024 * 1024;
+			MappedByteBuffer buffer =
+					channel.map(
+							FileChannel.MapMode.READ_ONLY,
+							0,
+							Math.min(channel.size(), Integer.MAX_VALUE));
+			buffer.order(byteOrder);
+			int bufferCount = 1;
+			// Java's NIO only allows memory-mapping up to 2GB. To work around this problem, we re-map
```
Comment (wko27) [inline src/main/java/com/medallia/word2vec/Word2VecModel.java:119]: Great comment :)

File: src/main/java/com/medallia/word2vec/Word2VecModel.java (hunk 1 above)
Comment (dirkgr) [inline src/main/java/com/medallia/word2vec/Word2VecModel.java]: I think Java will do an equivalent thing automatically for final constants like this, but I'm happy to follow your style guide :-)

File: src/main/java/com/medallia/word2vec/Word2VecModel.java (hunk 2 above)
Comment (dirkgr) [inline src/main/java/com/medallia/word2vec/Word2VecModel.java]: fixed

File: src/main/java/com/medallia/word2vec/Word2VecModel.java (hunk 4 above)
Comment (dirkgr) [inline src/main/java/com/medallia/word2vec/Word2VecModel.java]: I switched the whole thing over to using `ProfilingTimer`. I don't quite get what it's for. Normal Java logging frameworks already let you choose whether you want to see the output or not, and `ProfilingTimer` has no log levels.

**Comment by wko27** [comment]
Sorry for the delay!  Thank you for your work! :)

**Comment by dirkgr** [comment]
Yay!

**Comment by dirkgr** [comment]
@wko27, could I convince you to create a new release for this? We're still using a locally-built version, but we'd rather just depend on an "official" one from you.

**Comment by wko27** [comment]
Done! :)

**Comment by wko27** [comment]
Actually I might have misunderstood.  Did you want a maven release, or a Git release, or are the two tied together?  Sorry for the confusion ...

**Comment by dirkgr** [comment]
A Maven release is what I had in mind. But actually, I just finished another PR for this (https://github.com/medallia/Word2VecJava/pull/19). Maybe we do a review cycle of that, get it merged, and then you roll a release?

**Pull request merged**